repose-aggregator/core/repose-core/src/main/scala/org/openrepose/powerfilter/ReposeFilterLoader.scala

```diff
@@ -0,0 +1,321 @@
+    val duplicates = e.payload().asScala.groupBy(identity).collect { case (x, List(_, _, _*)) => x }
+    if (duplicates.isEmpty) {
+      healthCheckServiceProxy.resolveIssue(ApplicationDeploymentHealthReport)
+    } else {
+      val message = s"Please review your artifacts directory, multiple versions of the artifact${if (duplicates.size > 1) "s" else ""} exist! (${duplicates.mkString(",")})"
+      healthCheckServiceProxy.reportIssue(ApplicationDeploymentHealthReport, message, Severity.BROKEN)
+      logger.error(message)
+    }
+    if (artifactManager.allArtifactsLoaded) {
+      configurationCheck()
+    }
+  }
+
+  def setServletContext(newServletContext: ServletContext): Unit = {
+    currentServletContext = Option(newServletContext)
+    configurationCheck()
+  }
+
+  def getFilterContextList: Option[FilterContextList] = {
+    currentFilterContextRegistrar.map(_.bind())
+  }
+
+  def getTracingHeaderConfig: Option[TracingHeaderConfig] = currentSystemModel.map(_.getTracingHeader)
+
+  /**
+    * Triggered each time the event service triggers an app deploy and when the system model is updated.
+    */
+  private def configurationCheck(): Unit = {
+    (currentSystemModel, currentServletContext, artifactManager.allArtifactsLoaded) match {
+      case (Some(systemModel), Some(servletContext), true) =>
+        updateConfiguration(systemModel, servletContext)
+      case _ =>
+        logger.trace("{} -- Not ready to update yet.", nodeId)
+    }
+  }
+
+  private def updateConfiguration(systemModel: SystemModel, servletContext: ServletContext): Unit = {
+    configurationLock.synchronized {
+      val interrogator = new SystemModelInterrogator(nodeId)
+
+      Option(interrogator.getLocalCluster(systemModel).orElse(null)) match {
+        case Some(localCluster) =>
+          healthCheckServiceProxy.resolveIssue(SystemModelConfigHealthReport)
+          try {
+            // Only if we've been configured with some filters should we get a new list
+            // Sometimes we won't have any filters
+            val newFilterChain = Option(localCluster.getFilters) match {
+              case Some(listOfFilters: FilterList) => buildFilterContexts(servletContext, listOfFilters.getFilter.asScala.toList)
+              case _ => List.empty[FilterContext]
+            }
+
+            // Set the new FilterContextRegistrar and
+            // set the Close flag on the old one.
+            currentFilterContextRegistrar.synchronized {
+              val previousFilterContextRegistrar = currentFilterContextRegistrar
+              currentFilterContextRegistrar = Option(new FilterContextRegistrar(newFilterChain, Option(localCluster.getFilters.getBypassUriRegex)))
+              previousFilterContextRegistrar
+            }.foreach(_.close())
+
+            if (logger.underlying.isDebugEnabled) {
+              val filterChainInfo = newFilterChain.map(ctx => ctx.filter.getClass.getName).mkString(",")
```

Comment:
Is this really so expensive that we need this `if`?

Reply:
It has the potential to be expensive if there is a long Filter Chain. Since this is not the normal, I'm good not guarding it.
Update.